pallets/msa/src/lib.rs

```diff
@@ -738,3 +740,75 @@ impl<T: Config> AccountProvider for Pallet<T> {
 		Ok(result.unwrap())
 	}
 }
+
+/// Check to ensure that a provider has not yet been revoked if the
+/// calling extrinsic is revoking a provider to an msa.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct CheckProviderRevocation<T: Config + Send + Sync>(PhantomData<T>);
+
+impl<T: Config + Send + Sync> CheckProviderRevocation<T> {
+	/// Create new `SignedExtension` to check runtime version.
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Config + Send + Sync> sp_std::fmt::Debug for CheckProviderRevocation<T> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "CheckProviderRevocation<{:?}>", self.0)
+	}
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: Config + Send + Sync> SignedExtension for CheckProviderRevocation<T>
+where
+	T::Call: Dispatchable<Info = DispatchInfo> + IsSubType<Call<T>>,
+{
+	type AccountId = T::AccountId;
+	type Call = T::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "CheckProviderRevocation";
+
+	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		self.validate(who, call, info, len).map(|_| ())
+	}
+
+	fn validate(
+		&self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		match call.is_sub_type() {
+			Some(Call::revoke_msa_delegation_by_delegator { provider_msa_id, .. }) => {
```

Comment:
Looks like this will work, the only concern that I have is that someone flooding the queue with these requests and since the request has to do DB lookup(s) it would have effect of performance. One thing I was hoping we might be able to achieve was to group-by these events in queue in a way that you can not have more than one request per account at the same time, this would eliminate spamming using **one** account but does not help with using multiple different accounts to spam.

Reply:
we've talked about adding pre-checks to some of these calls before anything ever hits the message pool. I forgot what we decided about that, is this possible?

Reply:
I think I understand. You are worried that duplicate transactions flood the mempool right?

So while an invalid request would be stopped at the first honest node, a "valid" request could be duplicated and still pass this test?

Reply:
@wilwade yes

Reply:
I think we were talking about making this an unsigned transaction. I do not think this PR is ready.

Reply:
@enddynayn I thought we had rejected the idea of an unsigned transaction due to the tradeoff in complexity in getting the signature as a parameter setup.

I am curious as to the de-duplication more generally. I thought that was just testing nonces.

Reply:
This looks eerily similar 
https://hackernoon.com/zero-cost-blockchain-transactions-are-a-reality-on-polkadot

But looks like SignedExtension is way to go 

Are we sure we are not setting us up for vulnerabilities

Reply:
I added a prefix + tag unique to the origin of the revoke delegation call in the signed extension return value `ValidTransaction` of the validate function. This seems to affect the priority such that when a transaction with a tag is already queued, subsequent calls error out with a "Priority is too low" error. This might serve as a mechanism to prevent duplicate calls in the transaction queue.